# Define the Problem
[DataCamp Project: Predicting Credit Card Approvals](https://projects.datacamp.com/projects/558)

Commercial banks receive a lot of applications for credit cards. Many of them get rejected for many reasons, like high loan balances, low income levels, or too many inquiries on an individual's credit report, for example. Manually analyzing these applications is mundane, error-prone, and time-consuming (and time is money!). Luckily, this task can be automated with the power of machine learning and pretty much every commercial bank does so nowadays. In this notebook, we will build an automatic credit card approval predictor using machine learning techniques, just like the real banks do!

# Import/Collect the Data
We'll start by loading and viewing our data. We find that since this data is confidential, the contributor of the dataset has anonymized the feature names.

In [8]:
# Import pandas
import pandas as pd

lst_annon_cols = ['A1', 'A2', 'A3', 'A4', 'A5', 'A6', 'A7', 'A8', 
                   'A9', 'A10', 'A11', 'A12', 'A13', 'A14', 'A15', 'A16']

# Provided by http://rstudio-pubs-static.s3.amazonaws.com/73039_9946de135c0a49daa7a0a9eda4a67a72.html
lst_label_cols = ['Male','Age','Debt','Married','BankCustomer','EducationLevel',
                  'Ethnicity','YearsEmployed','PriorDefault','Employed','CreditScore',
                  'DriversLicense','Citizen','ZipCode','Income','Approved']

# Load dataset
df_cc_apps = pd.read_csv('http://archive.ics.uci.edu/ml/machine-learning-databases/credit-screening/crx.data',header=None)

# Add column names
df_cc_apps.columns = lst_label_cols
# Inspect data
df_cc_apps.head()

Male    Age   Debt Married BankCustomer EducationLevel Ethnicity  \
0    b  30.83  0.000       u            g              w         v   
1    a  58.67  4.460       u            g              q         h   
2    a  24.50  0.500       u            g              q         h   
3    b  27.83  1.540       u            g              w         v   
4    b  20.17  5.625       u            g              w         v   

   YearsEmployed PriorDefault Employed  CreditScore DriversLicense Citizen  \
0           1.25            t        t            1              f       g   
1           3.04            t        t            6              f       g   
2           1.50            t        f            0              f       g   
3           3.75            t        t            5              t       g   
4           1.71            t        f            0              f       s   

  ZipCode  Income Approved  
0   00202       0        +  
1   00043     560        +  
2   00280     824        +  
3   00100       3        +  
4   00120       0        +

# Explore the Data

In [16]:
# Print summary statistics
cc_apps_description = df_cc_apps.describe()
print(cc_apps_description)

             Debt  YearsEmployed  CreditScore         Income
count  690.000000     690.000000    690.00000     690.000000
mean     4.758725       2.223406      2.40000    1017.385507
std      4.978163       3.346513      4.86294    5210.102598
min      0.000000       0.000000      0.00000       0.000000
25%      1.000000       0.165000      0.00000       0.000000
50%      2.750000       1.000000      0.00000       5.000000
75%      7.207500       2.625000      3.00000     395.500000
max     28.000000      28.500000     67.00000  100000.000000


In [17]:
# Print DataFrame information
cc_apps_info = df_cc_apps.info()
print(cc_apps_info)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 690 entries, 0 to 689
Data columns (total 16 columns):
Male              690 non-null object
Age               690 non-null object
Debt              690 non-null float64
Married           690 non-null object
BankCustomer      690 non-null object
EducationLevel    690 non-null object
Ethnicity         690 non-null object
YearsEmployed     690 non-null float64
PriorDefault      690 non-null object
Employed          690 non-null object
CreditScore       690 non-null int64
DriversLicense    690 non-null object
Citizen           690 non-null object
ZipCode           690 non-null object
Income            690 non-null int64
Approved          690 non-null object
dtypes: float64(2), int64(2), object(12)
memory usage: 86.3+ KB
None


In [18]:
# Inspect missing values in the dataset
df_cc_apps.tail(10)

Male    Age    Debt Married BankCustomer EducationLevel Ethnicity  \
680    b  19.50   0.290       u            g              k         v   
681    b  27.83   1.000       y            p              d         h   
682    b  17.08   3.290       u            g              i         v   
683    b  36.42   0.750       y            p              d         v   
684    b  40.58   3.290       u            g              m         v   
685    b  21.08  10.085       y            p              e         h   
686    a  22.67   0.750       u            g              c         v   
687    a  25.25  13.500       y            p             ff        ff   
688    b  17.92   0.205       u            g             aa         v   
689    b  35.00   3.375       u            g              c         h   

     YearsEmployed PriorDefault Employed  CreditScore DriversLicense Citizen  \
680          0.290            f        f            0              f       g   
681          3.000            f        f            0              f       g   
682          0.335            f        f            0              t       g   
683          0.585            f        f            0              f       g   
684          3.500            f        f            0              t       s   
685          1.250            f        f            0              f       g   
686          2.000            f        t            2              t       g   
687          2.000            f        t            1              t       g   
688          0.040            f        f            0              f       g   
689          8.290            f        f            0              t       g   

    ZipCode  Income Approved  
680   00280     364        -  
681   00176     537        -  
682   00140       2        -  
683   00240       3        -  
684   00400       0        -  
685   00260       0        -  
686   00200     394        -  
687   00200       1        -  
688   00280     750        -  
689   00000       0        -

# Data Cleansing

## Handling Missing Values
We've uncovered some issues that will affect the performance of our machine learning model(s) if they go unchanged:

* Our dataset contains both numeric and non-numeric data (specifically data that are of float64, int64 and object types). Specifically, the features 2, 7, 10 and 14 contain numeric values (of types float64, float64, int64 and int64 respectively) and all the other features contain non-numeric values.
* The dataset also contains values from several ranges. Some features have a value range of 0 - 28, some have a range of 2 - 67, and some have a range of 1017 - 100000. Apart from these, we can get useful statistical information (like mean, max, and min) about the features that have numerical values.
* Finally, the dataset has missing values, which we'll take care of in this task. The missing values in the dataset are labeled with '?', which can be seen in the last cell's output.

Now, let's temporarily replace these missing value question marks with NaN.

In [23]:
# Import numpy
import numpy as np

# Inspect missing values in the dataset
print('Number of missing values in the dataset: ' + str(df_cc_apps.isnull().values.sum()))

Number of missing values in the dataset: 67


In [24]:
# Replace the '?'s with NaN
df_cc_apps = df_cc_apps.replace('?', np.NaN)

# Inspect the missing values again
df_cc_apps.tail(10)

Male    Age    Debt Married BankCustomer EducationLevel Ethnicity  \
680    b  19.50   0.290       u            g              k         v   
681    b  27.83   1.000       y            p              d         h   
682    b  17.08   3.290       u            g              i         v   
683    b  36.42   0.750       y            p              d         v   
684    b  40.58   3.290       u            g              m         v   
685    b  21.08  10.085       y            p              e         h   
686    a  22.67   0.750       u            g              c         v   
687    a  25.25  13.500       y            p             ff        ff   
688    b  17.92   0.205       u            g             aa         v   
689    b  35.00   3.375       u            g              c         h   

     YearsEmployed PriorDefault Employed  CreditScore DriversLicense Citizen  \
680          0.290            f        f            0              f       g   
681          3.000            f        f            0              f       g   
682          0.335            f        f            0              t       g   
683          0.585            f        f            0              f       g   
684          3.500            f        f            0              t       s   
685          1.250            f        f            0              f       g   
686          2.000            f        t            2              t       g   
687          2.000            f        t            1              t       g   
688          0.040            f        f            0              f       g   
689          8.290            f        f            0              t       g   

    ZipCode  Income Approved  
680   00280     364        -  
681   00176     537        -  
682   00140       2        -  
683   00240       3        -  
684   00400       0        -  
685   00260       0        -  
686   00200     394        -  
687   00200       1        -  
688   00280     750        -  
689   00000       0        -

We replaced all the question marks with NaNs. This is going to help us in the next missing value treatment that we are going to perform.

An important question that gets raised here is why are we giving so much importance to missing values? Can't they be just ignored? Ignoring missing values can affect the performance of a machine learning model heavily. While ignoring the missing values our machine learning model may miss out on information about the dataset that may be useful for its training. Then, there are many models which cannot handle missing values implicitly such as LDA.

So, to avoid this problem, we are going to impute the missing values with a strategy called mean imputation.

In [29]:
# Impute the missing values with mean imputation
df_cc_apps.fillna(df_cc_apps.mean(), inplace=True)

# Count the number of NaNs in the dataset to verify
print('number of NaNs in the dataset: ' + str(df_cc_apps.isnull().values.sum()))

number of NaNs in the dataset: 67
